File: app/src/main/java/de/westnordost/osmagent/dialogs/opening_hours/CircularSection.java
package de.westnordost.osmagent.dialogs.opening_hours;

import android.support.annotation.NonNull;
// TODO test

// actually closed-open: [start,end+1). ie.e Jun-Jul = 6 to 7 = both June and July
public class CircularSection implements Comparable<CircularSection>
{
	private final int start, end;

	public CircularSection(int start, int end)
	{
		this.start = start;
		this.end = end;
	}

	public int getStart()
	{
		return start;
	}

	public int getEnd()
	{
		return end;
	}

	public boolean intersects(CircularSection other)
	{
		if(loops() && other.loops())
			return true;
		if(loops() || other.loops())
			return other.end >= start || other.start <= end;
		return other.end >= start && other.start <= end;
	}

	public boolean loops()
	{
		return end < start;
	}

	@Override public int compareTo(@NonNull CircularSection other)
	{
		// loopers come first,
		if(loops() && !other.loops()) return -1;
		if(!loops() && other.loops()) return +1;
		// then by start
		int result = start - other.start;
		if(result != 0) return result;
		// then by end
		return end - other.end;
	}


}

File: app/src/main/java/de/westnordost/streetcomplete/osm/opening_hours/model/CircularSection.kt
package de.westnordost.streetcomplete.osm.opening_hours.model

import kotlinx.serialization.Serializable

/** An integer range that defines a section in a circle. The range that is defined is actually
 * closed-open: [start,end+1). i.e Jun-Jul (= start:6 end:7) shall be both June and July. If start
 * is bigger than end, it means that the section crosses the upper boundary. Think degrees.
 */
@Serializable
data class CircularSection(val start: Int, val end: Int) : Comparable<CircularSection> {

    fun intersects(other: CircularSection): Boolean =
        loops && other.loops ||
        if (loops || other.loops) {
            other.end >= start || other.start <= end
        } else {
            other.end >= start && other.start <= end
        }

    val loops get() = end < start

    override fun compareTo(other: CircularSection): Int {
        // loopers come first,
        if (loops && !other.loops) return -1
        if (!loops && other.loops) return +1
        // then by start
        val result = start - other.start
        if (result != 0) return result
        // then by end
        return end - other.end
    }

    override fun toString() = "$start-$end"
}
